DOCSP-46322: Raw queries (#143)
mongodb/docs-pymongo: source/interact-data/raw-queries.txt

.. _django-raw-queries:

============================
Perform Raw Database Queries
============================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: number, amount, estimation, code example

Overview
---------

In this guide, you can learn how to use {+django-odm+} to run
raw queries on your MongoDB database. Raw queries allow you to
query the database by using MongoDB's aggregation pipeline syntax
rather than Django methods.

The Django QuerySet API provides a ``QuerySet.raw()`` method, which allows

> you to perform raw SQL queries on relational databases. {+django-odm+}

Reviewer: S: To emphasize that the ODM is different
```suggestion
you to perform raw SQL queries on relational databases. However, {+django-odm+}
```

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``Movie`` and ``Theater`` models, which
represent collections in the ``sample_mflix`` database from the :atlas:`Atlas sample datasets </sample-data>`.
The model classes have the following definitions:

.. code-block:: python

    from django.db import models
    from django_mongodb_backend.fields import EmbeddedModelField, ArrayField
    from django_mongodb_backend.managers import MongoManager

    class Movie(models.Model):
        title = models.CharField(max_length=200)
        plot = models.TextField(null=True)
        runtime = models.IntegerField(default=0)
        released = models.DateTimeField("release date", null=True)
        awards = EmbeddedModelField(Award)
        genres = ArrayField(models.CharField(max_length=100), blank=True)
        objects = MongoManager()

        class Meta:
            db_table = "movies"

        def __str__(self):
            return self.title
    
    class Theater(models.Model):
        theaterId = models.IntegerField(default=0)
        objects = MongoManager()

        class Meta:
            db_table = "theaters"
        
        def __str__(self):

>             return self.title

Reviewer: I think this should probably be moved to an includes file so we can reuse it

Applies to all code examples

Author: Moving this to the [docs-django repo](https://github.com/mongodb/docs-django/pull/1) now - I'll put the code in a separate file there!

To learn how to create a Django application that uses a similar ``Movie``
model to interact with MongoDB documents, visit the :ref:`django-get-started`
tutorial.

.. _django-raw-queries-run:

Run Raw Queries
---------------

To run a raw database query, pass an aggregation pipeline
to the ``QuerySet.raw_aggregate()`` method. Aggregation pipelines
contain one or more stages that provide instructions on how to
process documents. After calling the ``raw_aggregate()`` method,
{+django-odm+} passes your pipeline to the ``pymongo.collection.Collection.aggregate()``
method and returns the query results as model objects.

.. tip::

   To learn more about constructing aggregation pipelines, see
   :manual:`Aggregation Pipeline </core/aggregation-pipeline/>`
   in the {+mdb-server+} manual.

This section shows how to use the ``raw_aggregate()`` method
to perform the following tasks:

- :ref:`django-raw-queries-filter-group`
- :ref:`django-raw-queries-search`
- :ref:`django-raw-queries-geospatial`

.. _django-raw-queries-filter-group:

Filter and Project Document Fields
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> This example runs a raw database query by calling the

Reviewer: ```suggestion
The following example runs a raw database query by calling the
```

Author: Since I use "following" in the sentence right before the code example (to introduce the pipeline stages used), I removed it from the first sentence to avoid repetition

        :visible: false

        Plot of The Parent Trap, released on 1961-06-21 00:00:00+00:00:
        Teenage twin girls swap places and scheme to reunite their divorced parents.

        Plot of The Parent Trap, released on 1998-07-29 00:00:00+00:00:
        Identical twins, separated at birth and each raised by one of their
        biological parents, discover each other for the first time at summer
        camp and make a plan to bring their wayward parents back together.

.. note::

   The ``raw_aggregate()`` method returns deferred model instances,
   which means that you can load fields omitted by the ``$project`` stage 
   on demand. In the preceding example, the query retrieves the ``title``
   and ``released`` fields. The print statement runs a separate query
   to retrieve the ``plot`` field.

.. _django-raw-queries-search:

Run an Atlas Search Query
~~~~~~~~~~~~~~~~~~~~~~~~~

You can run Atlas Search queries on your database to perform
fine-grained text searches. These queries provide advanced search
functionality, such as matching text phrases, scoring results for
relevance, and highlighting matches. 

To specify an Atlas Search query, create an Atlas Search index

> that covers the fields you want to query. Then, specify the ``$search``
> or ``$searchMeta`` stage in an aggregation pipeline parameter to
> the ``raw_aggregate()`` method.

Reviewer: ```suggestion
that covers the fields you want to query. Then, pass a ``$search``
or ``$searchMeta`` stage in an aggregation pipeline parameter to
the ``raw_aggregate()`` method.
```

.. important:: 

   You cannot use {+django-odm+} to create Atlas Search indexes.
   
   For instructions on using the PyMongo driver to create an Atlas
   Search index, see `Atlas Search and Vector Search Indexes
   <https://www.mongodb.com/docs/languages/python/pymongo-driver/current/indexes/atlas-search-index/>`__
   in the PyMongo documentation.
   
   For instructions on alternative methods of creating search indexes,
   see :atlas:`Create an Atlas Search Index </atlas-search/tutorial/create-index/>`
   in the Atlas documentation.

> This example runs an Atlas Search query by passing the ``$search`` pipeline

Reviewer: ```suggestion
The following example runs an Atlas Search query by passing the ``$search`` pipeline
```

Author: Same comment as above

        {'value': ' ', 'type': 'text'}, {'value': 'romance', 'type': 'hit'}, {'value': ' and embraces a new
        life as a thief, though he soon finds himself pursued by veteran police officer and engaged in a turf
        war with a local gangster.', 'type': 'text'}]}]

        Title: A Tale of Winter, text match details: [{'score': 3.3978850841522217, 'path': 'plot', 'texts':
        [{'value': 'Felicie and Charles have a serious if ', 'type': 'text'}, {'value': 'whirlwind', 'type':
        'hit'}, {'value': ' holiday ', 'type': 'text'}, {'value': 'romance', 'type': 'hit'}, {'value': '. ',
        'type': 'text'}]}]

.. important::

   When running the preceding example, ensure that you replace
   the ``<search-index-name>`` placeholder with the name of your
   Atlas Search index that covers the ``plot`` field.

.. _django-raw-queries-geospatial:

Query Geospatial Data
~~~~~~~~~~~~~~~~~~~~~

You can use the ``raw_aggregate()`` method to run queries
on fields containing geospatial data. Geospatial data represents
a geographic location on the surface of the Earth or on a
Euclidean plane.

To run a geospatial query, create a ``2d`` or ``2dsphere`` index on fields

> containing geospatial data. Then, specify one of the following

Reviewer: S: I think "pass" works a little better than "specify" in this context
```suggestion
containing geospatial data. Then, pass one of the following
```

query operators in an aggregation pipeline parameter to
the ``raw_aggregate()`` method:

- ``$near``
- ``$geoWithin``
- ``$nearSphere``
- ``$geoIntersects``

.. important:: 

   You cannot use {+django-odm+} to create ``2d`` or ``2dsphere`` indexes.
   
   For instructions on using the PyMongo driver to create geospatial
   indexes, see `Geospatial Indexes
   <https://www.mongodb.com/docs/languages/python/pymongo-driver/current/indexes/geospatial-index/>`__
   in the PyMongo documentation.
   
   For instructions on using the MongoDB Shell to create geospatial
   indexes, see :manual:`Geospatial Indexes </core/indexes/index-types/index-geospatial/>`
   in the {+mdb-server+} manual.

> This example runs a geospatial query by passing the ``$match`` and

Reviewer: ```suggestion
The following example runs a geospatial query by passing the ``$match`` and
```

Author: Same comment as above
